tests/test_prolongation.cpp

```diff
@@ -0,0 +1,293 @@
+    : public ::testing::TestWithParam<
+          std::tuple<int, bool>> //tuple includes data on grid size and Dirichlet boundary conditions
+{
+protected:
+    void SetUp() override
+    {
+        //initialize default parameters.
+        gyro::init_params();
+        gyro::icntl[Param::verbose]  = 0;
+        gyro::dcntl[Param::R0]       = 1e-5;
+        gyro::icntl[Param::periodic] = 1;
+        gyro::f_grid_r               = "";
+        gyro::f_grid_theta           = "";
+        gyro::f_sol_in               = "";
+        gyro::f_sol_out              = "";
+        gyro::icntl[Param::fac_ani]  = 3;
+        gyro::select_functions_class(gyro::icntl[Param::alpha_coeff], gyro::icntl[Param::beta_coeff],
+                                     gyro::icntl[Param::mod_pk], gyro::icntl[Param::prob]);
+    }
+};
+
+/*!
+ *  \brief Test the bilinear prolongation operator used in the multigrid cycle coarse-grid correction. 
+ *  
+ *  The Test creates an arbitrary grid-function on the coarser level and prolongates it. 
+ *  On the fine level we iterate over all nodes and test the result based on whether the node is fine in theta, r or in both.
+ *
+ *  Parametrized tests are used to test for different grid sizes and with or without Dirichlet boundary conditions.
+ */
+
+TEST_P(test_prolongation, test_bilinear_prolongation)
+{
+    //we vary the grid size to guarantee that the problem works for all sizes
+    gyro::icntl[Param::nr_exp]         = (int)(std::get<0>(GetParam()) / 3) + 3;
+    gyro::icntl[Param::ntheta_exp]     = (std::get<0>(GetParam()) % 3) + 3;
+    gyro::icntl[Param::DirBC_Interior] = std::get<1>(GetParam());
+    if (gyro::icntl[Param::nr_exp] == 3)
+        gyro::icntl[Param::fac_ani] = 2; //anisotropy should not exceed grid size
+
+    gmgpolar test_p;
+    create_grid(test_p);
+
+    level& p_level = *(test_p.v_level[0]);
+    int ctheta_int = test_p.v_level[1]->ntheta_int; //number of coarse nodes in theta direction
+
+    p_level.m  = test_p.v_level[0]->nr * test_p.v_level[0]->ntheta; //fine grid size
+    p_level.mc = test_p.v_level[1]->nr * test_p.v_level[1]->ntheta; //coarser grid size
+
+    std::vector<double> u_test(p_level.mc);
+    for (int z = 0; z < p_level.mc; z++) {
+        u_test[z] =
+            1 - z +
+            pow(PI,
+                -z * z); //constructing arbitrary grid-function on coarse level to test our prolongation operator with.
+    }
+
+    std::vector<double> sol =
+        p_level.apply_prolongation_bi(u_test); //apply prolongation operator on arbitrary grid-function
+
+    for (int j = 0; j < p_level.nr_int + 1; j++) {
+        for (int i = 0; i < p_level.ntheta_int; i++) {
+            if (j % 2 == 0 && i % 2 == 0) { //testing coarse node injection
+                EXPECT_EQ(u_test[(j / 2) * ctheta_int + (i / 2)], sol[j * p_level.ntheta_int + i])
+                    << "coarse node injection is failing";
+            }
+            else if (i % 2 != 0 && j % 2 == 0) { // theta_i is fine node
```

Comment:
```suggestion
            else if (i % 2 != 0 && j % 2 == 0) {
            // as numbering in angle (theta_i) is odd, we have a fine node with
            // coarse neighbors at (r_i, theta_j - k_{j-1}) and (r_i, theta_j + k_j)
```

Please adapt comment as following. theta_i itself cannot be coarse or fine. It is one coordinate of the node.

Reply:
i and j have to be exchanged